### ISSUE TITLE
Change Request: Track variables on SourceCode

### ISSUE THREAD

**Issue by nzakas** [eslint/css#160]
### Environment

HEAD

### What problem do you want to solve?

CSS variables are a significant part of modern CSS, and in order to accurately validate property values, each rule now has to track custom properties and see where they're applied in values.

### What do you think is the correct solution?

Something similar to the functionality of `eslint-scope` for CSS. I'm not quite sure on the details as the custom property/`var()` usage is a different paradigm than variables in JS, but something along the lines of a `sourceCode.customProperties` map where the key is the custom property name and the value is an object that tracks both rules where the custom property is set and declarations where the custom property value is read. We may also want to make it easy to get the closest value for that custom property to aid in validation, similar to what I did in https://github.com/eslint/css/pull/148.

### Participation

- [x] I am willing to submit a pull request for this change.

### Additional comments

We'd need an RFC for this to ensure we're thinking through the complexities.

**Comment by nzakas** [comment]
Ping @eslint/eslint-team

**Comment by lumirlumir** [comment]
The first time I saw it, I thought this subject was beyond my level, so I didn’t leave a comment.

But I personally support this suggestion. 👍

Variables hold a really important place in CSS.

**Comment by nzakas** [comment]
Ping @eslint/eslint-tsc

**Comment by mdjermanovic** [comment]
I'm not sure if it's feasible to track variables in css, for two reasons:

1. Custom properties can be defined in another file, or set in a script, so we generally can't know all possible values.
2. Even if we know all possible values, we still can't know which of them can apply to a particular `var()`.

It might be good to see the feedback when https://github.com/eslint/css/commit/9fb07fab74849e31ec363c75ce0d405a6a3108ec is released before proceeding.

**Comment by fasttime** [comment]
I think that a reasonable implementation for variable tracking will be necessarily oversimplified and miss a lot of edge cases, but I'm all in favor of this if there is a common need. In addition to `no-invalid-properties`, I imagine that a rule like `relative-font-units` could benefit from the ability to determine the possible values for a variable, e.g.:

```css
:root {
    .foo { --fs: 1rem; }
    .bar { --fs: 2em; }
}
div { font-size: var(--fs); }
```

> something along the lines of a `sourceCode.customProperties` map where the key is the custom property name and the value is an object that tracks both rules where the custom property is set and declarations where the custom property value is read.

Probably yes, it should be something like that at a minimum. Maybe it makes sense to start with a "dumb" data structure just to track how variables are defined and referenced in the current source, and see how that can be used to do smart things later, like getting the values for a variable in a selector context.

**Comment by JoshuaKGoldberg** [comment]
Are there any ideas on what could be feasibly achieved by tracking variables on SourceCode? I can't think of anything that isn't very broken by cross-file CSS:

* Unused property detection: they might be used in other files
* References to properties that don't exist: they might be declared in other files
* ...any others?

_Validation_ of properties (e.g. #148, enforcing conventions like naming or ordering) seems more realistic. But I can't think of any need there that needs more than just the declaration syntax.

**Comment by nzakas** [comment]
@JoshuaKGoldberg validation is the primary use case. Any rules that needs to understand what a variable's likely value is.

Cross-file is out of scope for this discussion.

**Comment by JoshuaKGoldberg** [comment]
I'm suggesting those rules should not be considered candidates because they're not feasible in production with the current constraints of ESLint.

**Comment by JoshuaKGoldberg** [comment]
Ok, now I'm getting lost myself 😄. Just to confirm, what are the two rules you're suggesting would need to know the value of variables?

> Candidates for what? 

Candidates for being added to `@eslint/css`. As in, rules that should be added in this repo, would need variable tracking, and therefore motivate this issue.

> And what does "feasible in production" mean?

I suspect the information available to ESLint within a single CSS file is not enough for rules to make informed decisions based on. As in, what was stated in  https://github.com/eslint/css/issues/160#issuecomment-2963992997 - I'd meant to agree with that comment.

I'm basing that off the assumption that many-to-most modern CSS projects don't store all their information in a single file. I think it's very common to, at the very least, put common variables in a separate file from what common components use. That file might be some shared CSS in the project or even from a shared design system.

Looking at the examples linked to or suggested in this thread, I think they're all very quickly broken by common cross-file CSS patterns...

**#148: validating property values containing variables**

Suppose a project has:

```scss
// base.css
:root {
  --color: red;
}
```

```scss
// some-component.css
.myComponent {
  color: var(--color);
}
```

That rule would erroneously report a lint error on `var(--color)`, right?

_(I would have mentioned this in the PR, but was on vacation at the time)_

**https://github.com/eslint/css/issues/160#issuecomment-2965183734: determining the possible values for a variable**

Suppose a project has:

```scss
// base.css
:root {
  --fs: 1rem;
}
```

```scss
// some-component.css
.someOverridesArea {
  --fs: 2rem;
}

.generalArea {
  font-size: var(--fs);
}
```

The lint rule would have no way of knowing whether:

* `.someOverridesArea` applies to everything targeted by `some-component.css`, or just some part of the component
* any other `--fs` value was set outside of the file

Thus, I don't think ESLint would be able to reliably say what `var(--fs)` may or may not be set to.

**#153 -> #174: feat: add `font-family-fallbacks` rule (draft)**

Suppose a project has:

```scss
// base.css
:root {
  --font-family-primary: "Open Sans";
  --font-family-secondary: sans-serif;
}
```

```scss
// some-component.css
.text {
  font-family: var(--font-family-primary) var(--font-family-secondary);
}
```

The tracker / lint rule would not be able to know what `var(--font-family-secondary)` is. Even if a rule like the `.someOverridesArea` in the last example were added, it wouldn't know whether `.text` is targeted by that.

**Comment by ayatobai** [comment]
@nzakas It looks like the RFC has been merged. If there aren’t any blockers, I can start implementing it.
